### PR TITLE
Ref alignment

### PR DESCRIPTION
**Pull request by TanmayGoyal**
Added a new folder for alignment, which has 4 classes defined, one base class (Align) and three child classes (PolyFit, LoessFit, ObiWarp).
The base class is Align, which is the parent class of all alignment algorithms.

Refactored background_peaks_update.cpp and alignmentdialog.cpp

PolyAligner class is now shifted inside PolyFit class. The files polyAligner.h and polyAligner.cpp is removed.

### CODE REVIEW

File: src/cli/peakdetector/peakdetector.pro [hunk 1]
```diff
@@ -14,7 +14,7 @@ QMAKE_CXXFLAGS += -std=c++11
 
 INCLUDEPATH +=  $$top_srcdir/src/core/libmaven  $$top_srcdir/3rdparty/pugixml/src $$top_srcdir/3rdparty/libneural $$top_srcdir/3rdparty/libpls \
 				$$top_srcdir/3rdparty/libcsvparser $$top_srcdir/3rdparty/libdate $$top_srcdir/3rdparty/libcdfread \
-				$$top_srcdir/src/pollyCLI $$top_srcdir/3rdparty/obiwarp
+				$$top_srcdir/src/pollyCLI $$top_srcdir/3rdparty/obiwarp $$top_srcdir/src/core/libmaven/alignment
```
Comment (sahil21) [inline src/cli/peakdetector/peakdetector.pro]: @TanmayGoyal No need to include the path of alignment folder.

File: src/core/libmaven/alignment/loessFit.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
+#include "mzAligner.h"
```
Comment (sahil21) [inline src/core/libmaven/alignment/loessFit.cpp]: @TanmayGoyal Don't include all header files in the cpp file. Only include loessFit.h

File: src/core/libmaven/alignment/polyFit.cpp [hunk 3]
```diff
@@ -1,7 +1,172 @@
-#include "PolyAligner.h"
+#include "mzAligner.h"
```
Comment (sahil21) [inline src/core/libmaven/alignment/polyFit.cpp]: Same thing here. Do not include all the header files here. Seems like a lot of unwanted header files are included here.

File: src/core/libmaven/alignment/polyFit.cpp [hunk 4]
```diff
@@ -1,7 +1,172 @@
...
+#include "polyFit.h"
+#include "mzMassSlicer.h"
+#include "mzSample.h"
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <QJsonArray>
```
Comment (sahil21) [inline src/core/libmaven/alignment/polyFit.cpp]: Removed unused header files please

File: tests/MavenTests/MavenTests.pro [hunk 5]
```diff
@@ -19,7 +19,7 @@ QMAKE_CXXFLAGS += -DOMP_PARALLEL
 
 INCLUDEPATH +=  $$top_srcdir/src/core/libmaven  $$top_srcdir/3rdparty/pugixml/src $$top_srcdir/3rdparty/libneural $$top_srcdir/3rdparty/libpls \
 				$$top_srcdir/3rdparty/libcsvparser $$top_srcdir/src/cli/peakdetector $$top_srcdir/3rdparty/libdate $$top_srcdir/3rdparty/libcdfread \
-                $$top_srcdir/3rdparty/obiwarp $$top_srcdir/src/pollyCLI
+                $$top_srcdir/3rdparty/obiwarp $$top_srcdir/src/pollyCLI $$top_srcdir/src/core/libmaven/alignment
```
Comment (sahil21) [inline tests/MavenTests/MavenTests.pro]: @TanmayGoyal No need to include this

File: src/gui/mzroll/tabledockwidget.h [hunk 6]
```diff
@@ -10,7 +10,11 @@
 #include "numeric_treewidgetitem.h"
 #include "QHistogramSlider.h"
 #include "saveJson.h"
-#include "jsonReports.h";
+#include "jsonReports.h"
+#include "polyFit.h"
```
Comment (sahil21) [inline src/gui/mzroll/tabledockwidget.h]: @TanmayGoyal Why do we need to include alignment in tabledockwidget?

File: src/gui/mzroll/mzfileio.h [hunk 7]
```diff
@@ -4,8 +4,9 @@
 #include <mutex>
 #include "globals.h"
 #include "mainwindow.h"
-#include "mzAligner.h"
-
+#include "polyFit.h"
```
Comment (sahil21) [inline src/gui/mzroll/mzfileio.h]: @TanmayGoyal seems like unused header files

File: src/gui/mzroll/mainwindow.h [hunk 8]
```diff
@@ -344,8 +347,8 @@ public Q_SLOTS:
 	void setUrl(Compound*);
 	void setUrl(Reaction*);
 	void setFormulaFocus(QString formula);
-	void Align();
-	void UndoAlignment();
+	// void Align();
```
Comment (sahil21) [inline src/gui/mzroll/mainwindow.h]: @TanmayGoyal Remove these commented code

File: src/cli/peakdetector/peakdetector.pro (hunk 1 above)
Comment (TanmayGoyal) [reply src/cli/peakdetector/peakdetector.pro]: Done, removed alignment folder path from all project files

File: src/gui/mzroll/tabledockwidget.h (hunk 6 above)
Comment (TanmayGoyal) [reply src/gui/mzroll/tabledockwidget.h]: @sahil21 We need to include one of the header files for alignment algorithms ( which is currently polyfit ) as tabledockwidget uses it. We can however remove the other 2 header files.

File: src/core/libmaven/PeakDetector.cpp [hunk 9]
```diff
@@ -267,13 +267,17 @@ void PeakDetector::alignSamples() {
                 for (unsigned int i = 0; i < mavenParameters->allgroups.size(); i++)
                         agroups[i] = &mavenParameters->allgroups[i];
                 //init aligner
-                Aligner aligner;
-                aligner.doAlignment(agroups);
+                // Aligner aligner;
+                // aligner.doAlignment(agroups);
+
+                PolyFit *polyFit = new PolyFit(agroups, mavenParameters->samples);
```
Comment (sahil21) [inline src/core/libmaven/PeakDetector.cpp]: @TanmayGoyal Why are we passing samples here? Don't think that is required

File: src/core/libmaven/PeakDetector.cpp [hunk 10]
```diff
@@ -267,13 +267,17 @@ void PeakDetector::alignSamples() {
                 for (unsigned int i = 0; i < mavenParameters->allgroups.size(); i++)
                         agroups[i] = &mavenParameters->allgroups[i];
                 //init aligner
-                Aligner aligner;
-                aligner.doAlignment(agroups);
+                // Aligner aligner;
```
Comment (shubhra-agrawal) [inline src/core/libmaven/PeakDetector.cpp]: @TanmayGoyal Comments should be removed

File: src/core/libmaven/PeakDetector.cpp [hunk 11]
```diff
@@ -267,13 +267,17 @@ void PeakDetector::alignSamples() {
...
+
+                PolyFit *polyFit = new PolyFit(agroups);
+                polyFit->polyFitAlgo();
                 mavenParameters->writeCSVFlag = true;
         }
 }
 
-void PeakDetector::processSlices(vector<mzSlice *> &slices, string setName)
+void PeakDetector::processSlices(vector<mzSlice *> &slices, string setName) 
+
```
Comment (shubhra-agrawal) [inline src/core/libmaven/PeakDetector.cpp]: Extra line

**Pull request closed without merging**